File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.java/core/src/test/java/de/unistuttgart/hamster/InitHamsterCommandTest.java
package de.unistuttgart.hamster;

import de.unistuttgart.hamster.framework.CommandConstraintException;
import de.unistuttgart.hamster.hamster.HamsterGame;
import de.unistuttgart.hamster.mpw.Direction;
import de.unistuttgart.hamster.mpw.Location;
import de.unistuttgart.hamster.util.GameStringifier;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class InitHamsterCommandTest {
	
	@Test
	public void testInitHamsterOn1x1_and4Grains() {
		HamsterGame game = GameStringifier.createFromString("   ;" +
															"   ;");

		var sut = game.getTerritory().getDefaultHamster();
		sut.init(game.getTerritory(), locationOf(1, 1), Direction.SOUTH, 4);

		String actual = GameStringifier.toString(game);
		assertEquals("   ;" +
				     " v ;", actual);
		assertEquals(4, sut.getInternalHamster().getGrains().size());
	}


	@Test
	public void testInitHamster_invalidGrainCount() {
		HamsterGame game = GameStringifier.createFromString(" ;");

		var sut = game.getTerritory().getDefaultHamster();
		assertThrows(CommandConstraintException.class, () -> {
			sut.init(game.getTerritory(), locationOf(0, 0), Direction.SOUTH, -1);
		});
	}

	@Test
	public void testUndoInitHamster() {
		HamsterGame game = GameStringifier.createFromString("  ;");

		var sut = game.getTerritory().getDefaultHamster();
		sut.init(game.getTerritory(), locationOf(0, 0), Direction.SOUTH, 2);
		game.getGameCommandStack().undo();

		assertEquals(null, sut.getInternalHamster().getDirection());
		assertEquals(0, sut.getInternalHamster().getGrains().size());
	}

	private Location locationOf(int column, int row) {
		var location = new Location();
		location.setColumn(column);
		location.setRow(row);
		return location;
	}


}

File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/test/InitHamsterCommandTest.cpp
#include "gtest/gtest.h"

#include "HamsterCore.h"

#include "util/GameStringifier.h"
#include "util/TestUtils.h"

#include <utility>
#include <vector>
#include <string>

using namespace mpw;
using namespace hamster;
using namespace framework;
using namespace util;

/// \note use 'NO-LINT' comment at tests to suppress a warning caused by TEST_F
class InitHamsterCommandTest : public testing::Test {
private:
public:
    static Location locationOf(int column, int row);
};

TEST_F(InitHamsterCommandTest, testInitHamsterOn1x1_and4Grains) { /* NOLINT */
    auto game = GameStringifier::createFromString("   ;"
                                                        "   ;");

    auto sut = game->getTerritory()->getDefaultHamster();
    sut->init(game->getTerritory(), locationOf(1, 1), Direction::SOUTH, 4);

    std::string actual = GameStringifier::toString(*game);
    EXPECT_EQ("   ;"
                 " v ;", actual);
    EXPECT_EQ(4, sut->getInternalHamster()->getGrains().size());
}

TEST_F(InitHamsterCommandTest, testInitHamster_invalidGrainCount) { /* NOLINT */
    auto game = GameStringifier::createFromString(" ;");

    auto sut = game->getTerritory()->getDefaultHamster();
    TestUtils::assertThrows(typeid(CommandConstraintException), [&]() {
        sut->init(game->getTerritory(), locationOf(0, 0), Direction::SOUTH, -1);
    });
}

TEST_F(InitHamsterCommandTest, testUndoInitHamster) { /* NOLINT */
    auto game = GameStringifier::createFromString("  ;");

    auto sut = game->getTerritory()->getDefaultHamster();
    sut->init(game->getTerritory(), locationOf(0, 0), Direction::SOUTH, 2);
    game->getCommandStack()->undo();

    EXPECT_EQ(Direction::NORTH, sut->getInternalHamster()->getDirection());
    EXPECT_EQ(0, sut->getInternalHamster()->getGrains().size());
}

//<editor-fold desc="helpers">

Location InitHamsterCommandTest::locationOf(int column, int row) {
    return Location::from(column, row);
}

//</editor-fold>
